packages/block-editor/src/components/media-flow/index.js

```diff
@@ -0,0 +1,186 @@
+	{
+		mediaUpload,
+		mediaURL,
+		allowedTypes,
+		onSelect,
+		onSelectURL,
+		noticeOperations,
+		name = __( 'Replace' ),
+		multiple = false,
+	}
+) => {
+	const [ showURLInput, setshowURLInput ] = useState( false );
+	const [ showEditURLInput, setshowEditURLInput ] = useState( false );
+	const [ mediaURLValue, setMediaURLValue ] = useState( mediaURL );
+
+	let mediaFlowOptionsMenu;
+	const getMediaFlowRef = ( element ) => {
+		mediaFlowOptionsMenu = element;
+	};
+
+	const selectMedia = ( media ) => {
+		onSelect( media );
+		setMediaURLValue( media.url );
+	};
+
+	const onUploadError = ( message ) => {
+		noticeOperations.removeAllNotices();
+		noticeOperations.createErrorNotice( message );
+	};
+
+	const uploadFiles = ( event ) => {
+		const files = event.target.files;
+		let setMedia;
+		if ( multiple ) {
+			setMedia = selectMedia;
+		} else {
+			setMedia = ( [ media ] ) => {
+				if ( !! mediaFlowOptionsMenu ) {
+					mediaFlowOptionsMenu.close();
+				}
+				selectMedia( media );
+			};
+		}
+		mediaUpload( {
+			allowedTypes,
+			filesList: files,
+			onFileChange: setMedia,
+			onUploadError,
+		} );
+	};
+
+	const fileUploadButton = (
+		<FormFileUpload
+			onChange={ uploadFiles }
+			accept={ allowedTypes }
+			multiple={ multiple }
+			render={ ( { openFileDialog } ) => {
+				return (
+					<MenuItem
+						icon="upload"
+						onClick={ openFileDialog }
+					>
+						{ __( 'Upload' ) }
+					</MenuItem>
+				);
+			} }
+		/>
+	);
+
+	const URLButton = (
+		<>
+			<MenuItem
+				icon="admin-links"
+				onClick={ () => ( setshowURLInput( ! showURLInput ) ) }
+			>
+				<div> { __( 'Insert from URL' ) } </div>
+			</MenuItem>
+		</>
+	);
+
+	const urlInputUI = ( showURLInput && <div className="block-editor-media-flow__url-input">
+		{ showEditURLInput ? ( <LinkEditor
+			value={ mediaURLValue }
+			isFullWidth={ true }
+			hasBorder={ true }
+			onChangeInputValue={ ( url ) => ( setMediaURLValue( url ) ) }
+			onSubmit={ ( event ) => {
+				event.preventDefault();
+				onSelectURL( mediaURLValue );
+				setshowEditURLInput( ! showEditURLInput );
+			} }
+		/> ) : ( <LinkViewer
+			isFullWidth={ true }
+			className="editor-format-toolbar__link-container-content block-editor-format-toolbar__link-container-content"
+			url={ mediaURLValue }
+			onEditLinkClick={ () => ( setshowEditURLInput( ! showEditURLInput ) ) }
+		/> )
+		} </div> );
+
+	const editMediaButton = (
+		<BlockControls>
+			<MediaUploadCheck>
+				<MediaUpload
+					onSelect={ ( media ) => selectMedia( media ) }
+					onClose={ () => {
+						mediaFlowOptionsMenu.close();
+					} }
+					allowedTypes={ allowedTypes }
+					render={ ( { open } ) => (
+						<>
+							<Toolbar
+								isCollapsed={ true }
+								dropDownRef={ getMediaFlowRef }
+								icon={ false }
+								label={ name }
+								showLabel={ true }
+								className={ 'media-flow_toolbar' }
+								onToggle={ () => ( setshowURLInput( false ) ) }
+								helperUI={ urlInputUI }
```

Comment:
I'm not sure I understand why do we need these new props `onToggle` and `helperUI`. These look specific to me to the URLButton menu item and would probably be better handled together?

Reply:
We are not using an urlButton. The helperUI is a way to add more UI in a dropdown menu which are not menu items. onToggle is needed if there is state to be cleared inside the dropdown menu.